components/provisioner/internal/model/gardener_config.go

```diff
@@ -356,20 +314,82 @@ func (c AWSGardenerConfig) ExtendShootConfig(gardenerConfig GardenerConfig, shoo
+	ProviderSpecificConfig
+	input *gqlschema.OpenStackProviderConfigInput `db:"-"`
+}
+
+func NewOpenStackGardenerConfig(input *gqlschema.OpenStackProviderConfigInput) (*OpenStackGardenerConfig, apperrors.AppError) {
+	config, err := json.Marshal(input)
+	if err != nil {
+		return &OpenStackGardenerConfig{}, apperrors.Internal("failed to marshal OpenStack Gardener config")
+	}
+
+	return &OpenStackGardenerConfig{
+		ProviderSpecificConfig: ProviderSpecificConfig(config),
+		input:                  input,
+	}, nil
+}
+
+func (c OpenStackGardenerConfig) AsProviderSpecificConfig() gqlschema.ProviderSpecificConfig {
+	return gqlschema.OpenStackProviderConfig{
+		Zones:                c.input.Zones,
+		FloatingPoolName:     c.input.FloatingPoolName,
+		CloudProfileName:     c.input.CloudProfileName,
+		LoadBalancerProvider: c.input.LoadBalancerProvider,
+	}
+}
+
+func (c OpenStackGardenerConfig) EditShootConfig(gardenerConfig GardenerConfig, shoot *gardener_types.Shoot) apperrors.AppError {
+	return updateShootConfig(gardenerConfig, shoot, c.input.Zones)
+}
+
+func (c OpenStackGardenerConfig) ExtendShootConfig(gardenerConfig GardenerConfig, shoot *gardener_types.Shoot) apperrors.AppError {
+	shoot.Spec.CloudProfileName = c.input.CloudProfileName
+
+	workers := []gardener_types.Worker{getWorkerConfig(gardenerConfig, c.input.Zones)}
+
+	openStackInfra := NewOpenStackInfrastructure(c.input.FloatingPoolName, gardenerConfig.WorkerCidr)
+	jsonData, err := json.Marshal(openStackInfra)
+	if err != nil {
+		return apperrors.Internal("error encoding infrastructure config: %s", err.Error())
+	}
+
+	openstackControlPlane := NewOpenStackControlPlane(c.input.LoadBalancerProvider)
+	jsonCPData, err := json.Marshal(openstackControlPlane)
+	if err != nil {
+		return apperrors.Internal("error encoding control plane config: %s", err.Error())
+	}
+
+	shoot.Spec.Provider = gardener_types.Provider{
+		Type:                 "openstack",
+		ControlPlaneConfig:   &apimachineryRuntime.RawExtension{Raw: jsonCPData},
+		InfrastructureConfig: &apimachineryRuntime.RawExtension{Raw: jsonData},
+		Workers:              workers,
+	}
+
+	return nil
+}
+
 func getWorkerConfig(gardenerConfig GardenerConfig, zones []string) gardener_types.Worker {
-	return gardener_types.Worker{
+	worker := gardener_types.Worker{
 		Name:           "cpu-worker-0",
 		MaxSurge:       util.IntOrStringPtr(intstr.FromInt(gardenerConfig.MaxSurge)),
 		MaxUnavailable: util.IntOrStringPtr(intstr.FromInt(gardenerConfig.MaxUnavailable)),
 		Machine:        getMachineConfig(gardenerConfig),
-		Volume: &gardener_types.Volume{
-			Type:       &gardenerConfig.DiskType,
-			VolumeSize: fmt.Sprintf("%dGi", gardenerConfig.VolumeSizeGB),
-		},
-		Maximum: int32(gardenerConfig.AutoScalerMax),
-		Minimum: int32(gardenerConfig.AutoScalerMin),
-		Zones:   zones,
+		Maximum:        int32(gardenerConfig.AutoScalerMax),
+		Minimum:        int32(gardenerConfig.AutoScalerMin),
+		Zones:          zones,
 	}
+
```

Comment:
I don't really remember where could it be but do you think it'll be useful to validate the graphql input with the optional `DiskType` and `VolumeSizeGB` fields? I assume that the provisioning and upgrading of the shoot will just fail if they regard the OpenStack clusters and these two fields are provided. I think we have someplace to validate the input so we could fail these requests super fast

Reply:
Thanks!